clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala

```diff
@@ -227,6 +236,103 @@ object GarbageCollector {
                                                spark,
                                                APIConfigurations(apiURL, accessKey, secretKey)
                                               ).withColumn("run_id", lit(runID))
-    expiredAddresses.write.partitionBy("run_id").mode(SaveMode.Append).parquet(addressesDFLocation)
+    expiredAddresses.write
+      .partitionBy("run_id")
+      .mode(SaveMode.Append)
+      .parquet(addressesDFLocation) // TODO(Guys): consider changing to overwrite
+  }
+}
+
+object S3BulkDeleter {
+  def repartitionBySize(df: DataFrame, maxSize: Int, column: String): DataFrame = {
+    val nRows = df.count()
+    val nPartitions = math.max(1, math.floor(nRows / maxSize)).toInt
+    df.repartitionByRange(nPartitions, col(column))
+  }
+
+  def delObjIteration(bucket: String, keys: Seq[String], s3Client: S3Client): Seq[String] = {
+    if (keys.isEmpty) None
+    val removeKeys = keys.map(ObjectIdentifier.builder().key(_).build()).asJava
+    val delObj = Delete.builder().objects(removeKeys).build()
+    val delObjReq = DeleteObjectsRequest.builder.delete(delObj).bucket(bucket).build()
+    val res = s3Client.deleteObjects(delObjReq)
+    res.deleted().asScala.map(_.key())
+  }
+
+  private def getS3Client(region: String, numRetries: Int) = {
+    val retryPolicy = RetryPolicy
+      .builder()
+      .backoffStrategy(BackoffStrategy.defaultThrottlingStrategy())
+      .numRetries(numRetries)
+      .build() // TODO: change to configurable
+    val configuration = ClientOverrideConfiguration.builder().retryPolicy(retryPolicy).build()
+    S3Client.builder.region(Region.of(region)).overrideConfiguration(configuration).build
+  }
+
+  def bulkRemoveFromIter(
+      keys: Iterator[String],
+      bucket: String,
+      region: String,
+      bulkSize: Int,
+      numRetries: Int
+  ): Iterator[mutable.Buffer[String]] = {
+    var nextBatch = keys.take(bulkSize)
+    var res = Seq[mutable.Buffer[String]]()
+    val s3Client = getS3Client(region, numRetries)
```

Comment:
Could we create the client once and pass that instead of passing `region` and `numRetries` parameters that are constant for the entire bulk remove operation?

Reply:
No, because it's not serializable